File: typescript/src/problems201_300/reverse-linked-list.ts
import SinglyListNode from '../concepts/singly-list-node';

class Problem {
  reverseList(head) {
    throw new Error(`Not implemented (${head}).`);
  }
}

class Default extends Problem {
  reverseList(head) {
    let previous: SinglyListNode | undefined = undefined;
    let current: SinglyListNode | undefined = head;
    while (current) {
      const next: SinglyListNode | undefined = current.next;
      current.next = previous;
      previous = current;
      current = next;
    }
    return previous;
  }
}

const SOLUTIONS = [new Default()];
export default SOLUTIONS;

File: python/src/problems201_300/reverse_linked_list.py
from abc import ABC, abstractmethod
from typing import override

from src.concepts.singly_list_node import SinglyListNode


class Problem(ABC):
    @abstractmethod
    def reverse_list(self, head: SinglyListNode | None) -> SinglyListNode | None:
        pass


class Default(Problem):
    @override
    def reverse_list(self, head: SinglyListNode | None) -> SinglyListNode | None:
        previous: SinglyListNode | None = None
        current: SinglyListNode | None = head
        while current:
            next_node: SinglyListNode | None = current.next
            current.next = previous
            previous = current
            current = next_node
        return previous


SOLUTIONS: list[Problem] = [Default()]
